include/init.h

```diff
@@ -29,49 +30,101 @@ extern "C" {
 #define _SYS_INIT_LEVEL_SMP		4
 #endif
 
+struct device;
+
+/**
+ * @brief Static init entry structure for each device driver or services
+ *
+ * @param dev pointer to a device driver instance structure. Can be NULL
+ * if this init entry is not used for a device driver.
+ * @param init init function for the init entry. Also used to initialize
+ * the device driver instance represented by
+ */
+struct init_entry {
+	int (*init)(struct device *dev);
```

Comment:
The review version:  We need to decide, and express clearly, whether there can be init functions that take a non-null argument that points to something other than a device.  If so, then the parameter declaration should be `void * arg` to avoid forced MISRA violations.

If the null/non-null status of the `dev` field is sufficient to distinguish device initialization from all other initializations then the types don't need to change, but the documentation should be "~Can~ Shall be NULL if and only if this....".

Reply:
The question we need to ask ourselves here: are we going to initialize objects - other than struct device - that could exist as n instances for the same initialization function?

Reply:
Agreed.  An outdated example might be all the pimux-based init functions: rather than hard-coding them into a single function sprinkled with `#ifdef` regions, the settings might be recorded in per-instance tables, and a SOC-specific init function used to process the settings for each instance that is actually linked in.

I *think* I could accept an argument that any case where an init function might be shared among multiple clients could be handled by a level of indirection: register a wrapper function with no arguments, which invokes a subordinate function on each of the objects it's supposed to affect.  This might waste space when multiple wrappers are needed to enforce dependency order.